File: files/CH-167.py
import pandas as pd
import numpy as np
import re

path = "CH-167 Table Transformation.xlsx"
input = pd.read_excel(path, usecols="C", skiprows=1, nrows=25)
test = pd.read_excel(path, usecols="E:G", skiprows=1, nrows=10)

def classify_type(value):
    str_val = str(value)
    if re.match(r'^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}$', str_val):
        return 'date'
    if re.match(r'^[A-Za-z]+$', str_val):
        return 'letter'
    if re.match(r'^\d+$', str_val):
        return 'digit'
    return 'unknown'

input['type'] = input.iloc[:, 0].apply(classify_type)
input['group'] = (input['type'] == 'date').cumsum()

result = input.groupby(['group', 'type'])[input.columns[0]].apply(list).unstack().reset_index()
result = result.explode(['letter', 'digit']).reset_index(drop=True)
result.columns.name = None

result = result[['date', 'letter', 'digit']]
result.columns = ['Date', 'Product', 'Quantity']

result['Date'] = pd.to_datetime(result['Date'].apply(lambda x: x[0] if isinstance(x, list) else x), errors='coerce')
result['Quantity'] = pd.to_numeric(result['Quantity'], errors='coerce').astype('int64')

print(result.equals(test)) # True

File: files/CH-167.R
library(tidyverse)
library(readxl)

path = "files/CH-167 Table Transformation.xlsx"
input = read_excel(path, range = "C2:C27")
test  = read_excel(path, range = "E2:G12")

result = tibble(raw = input$`Column 1`) %>%
  mutate(
    type = case_when(
      str_detect(raw, "^\\d{5}$") ~ "date",
      str_detect(raw, "^[A-Za-z]+$") ~ "letter",
      str_detect(raw, "^\\d+$") ~ "digit",
      TRUE ~ "unknown"
    ),
    group = cumsum(type == "date")
  ) %>%
  pivot_wider(names_from = type, values_from = raw, values_fn = list(raw = list)) %>%
  unnest(cols = c(date, letter, digit)) %>%
  select(Date = date, Product = letter, Quantity = digit) %>%
  mutate(Date = as.POSIXct(as.Date(as.numeric(Date), origin = "1899-12-30")),
         Quantity = as.numeric(Quantity))

all.equal(result, test, check.attributes = FALSE)
#> [1] TRUE